metronome/core/test/src/io/iohk/metronome/core/messages/RPCTrackerSpec.scala

```diff
@@ -0,0 +1,78 @@
+package io.iohk.metronome.core.messages
+
+import monix.eval.Task
+import monix.execution.Scheduler.Implicits.global
+import org.scalatest.flatspec.AsyncFlatSpec
+import org.scalatest.matchers.should.Matchers
+import org.scalatest.compatible.Assertion
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+class RPCTrackerSpec extends AsyncFlatSpec with Matchers {
+
+  sealed trait TestMessage extends RPCMessage
+  object TestMessage extends RPCMessageCompanion {
+    case class FooRequest(requestId: RequestId) extends TestMessage with Request
+    case class FooResponse(requestId: RequestId, value: Int)
+        extends TestMessage
+        with Response
+    case class BarRequest(requestId: RequestId) extends TestMessage with Request
+    case class BarResponse(requestId: RequestId, value: String)
+        extends TestMessage
+        with Response
+
+    implicit val foo = pair[FooRequest, FooResponse]
+    implicit val bar = pair[BarRequest, BarResponse]
+  }
+  import TestMessage._
+
+  def test(
+      f: RPCTracker[Task, TestMessage] => Task[Assertion]
+  ): Future[Assertion] =
+    RPCTracker[Task, TestMessage](10.seconds)
+      .flatMap(f)
+      .timeout(5.seconds)
+      .runToFuture
+
+  behavior of "RPCTracker"
+
+  it should "complete responses within the timeout" in test { tracker =>
+    val req = FooRequest(RequestId())
+    val res = FooResponse(req.requestId, 1)
+    for {
+      join <- tracker.register(req)
+      ok   <- tracker.complete(res)
+      got  <- join
+    } yield {
+      ok shouldBe true
+      got shouldBe Some(res)
+    }
+  }
+
+  it should "complete responses with None after the timeout" in test {
+    tracker =>
+      val req = FooRequest(RequestId())
+      for {
+        join <- tracker.register(req, timeout = 50.millis)
+        _    <- Task.sleep(100.millis)
+        got  <- join
+      } yield {
+        got shouldBe empty
+      }
+  }
+
+  it should "complete responses with None if the wrong type of response arrives" in test {
+    tracker =>
+      val req = FooRequest(RequestId())
+      val res = BarResponse(RequestId(), "one")
```

Comment:
If I understand correctly the failed `complete` is not really due to the wrong type but due to a different UUID. In fact, if the Ids matched the request would be completed with the wrong response type, would it not?

Reply:
Good point about the request ID, but it wouldn't be completed with the wrong type because of the `ClassTag` which is passed to `register`, it's used to only match responses of the expected type.

Reply:
Fixed the test. It turns out the code didn't do what the test expected, it allowed the timeout to happen rather than complete the operation with `None`. I thought which one's the better behaviour, but fixed it so `None` is returned immediately because:
* A malicious user can try to stall us by just not sending anything, they have no incentive to send the wrong kind of data.
* If we get the wrong kind of data it's most likely down to some programming error, very unlikely that a correct response will follow.
* Maybe it's an error with a certain node, and returning immediately gives us the option to move on and ask somebody else.

But I may be wrong, maybe allowing it to time out and a potentially correct message to arrive is better.

Reply:
I also prefer returning `None`. It should be easier to catch any errors and I don't think it can be abused